### ISSUE TITLE
Disable autocast simple decorator

### ISSUE THREAD

**Issue by timurlenk07** [pytorch/pytorch#72421]
### 🚀 The feature, motivation and pitch

I was working on reimplementing the YOLOX object detection method, with planned support for mixed prec, but in a part of it, I had to calculate binary crossentropy of a tensor, that itself was a product of two other tensors with sigmoid applied.
```python
...
cls_preds = (cls_preds.sigmoid() * obj_preds.sigmoid()).sqrt()
pair_wise_cls_loss = F.binary_cross_entropy(cls_preds, gt_cls_per_image, reduction="none")
...
```
Since `binary_cross_entropy` is numerically unstable, when using mixed precision the training exits with a RuntimeError. To get around this problem, I wanted to locally disable autocast, just for these lines. It turned out that I had to specify the device to use the `torch.autocast()` function. When specifying `'cuda'` as the device, in case I was on GPU, it obviously worked as expected, but the funny thing was, that when the tensors were on *cpu*, it only showed a warning (a single time) then the training proceeded as expected.

So my question is, is it possible to add a *device agnostic* call to disable native mixed precision of PyTorch? It would make sense to me, since `device` is only needed to determine the dtype to use during mixed prec calculations, but as far as I know, `float32` is the default dtype for all devices.

### Alternatives

The alternative I tried (and worked, with a single warning):
```python
with autocast('cuda', enabled=False):
    cls_preds = (cls_preds.sigmoid() * obj_preds.sigmoid()).sqrt()
    pair_wise_cls_loss = F.binary_cross_entropy(cls_preds, gt_cls_per_image, reduction="none")
```

### Additional context

I searched for similar issues but didn't manage to find anything. Sorry if I missed something 🤷
